services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/MappingAndDelegatingCommandHandler.java

```diff
@@ -0,0 +1,156 @@
+    private final CommandTargetMapper commandTargetMapper;
+    private final KafkaBasedInternalCommandSender internalCommandSender;
+    private final Tracer tracer;
+
+    /**
+     * Creates a new MappingAndDelegatingCommandHandler instance.
+     *
+     * @param commandTargetMapper The mapper component to determine the command target.
+     * @param internalCommandSender The command sender to publish commands to the internal command topic.
+     * @param tracer The tracer instance.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public MappingAndDelegatingCommandHandler(
+            final CommandTargetMapper commandTargetMapper,
+            final KafkaBasedInternalCommandSender internalCommandSender,
+            final Tracer tracer) {
+        Objects.requireNonNull(commandTargetMapper);
+        Objects.requireNonNull(internalCommandSender);
+        Objects.requireNonNull(tracer);
+
+        this.commandTargetMapper = commandTargetMapper;
+        this.internalCommandSender = internalCommandSender;
+        this.tracer = tracer;
+    }
+
+    @Override
+    public Future<Void> start() {
+        return internalCommandSender.start();
+    }
+
+    @Override
+        public Future<Void> stop() {
+        return internalCommandSender.stop();
+    }
+
+    /**
+     * Delegates an incoming command to the protocol adapter instance that the target
+     * device is connected to.
+     * <p>
+     * Determines the target gateway (if applicable) and protocol adapter instance for an incoming command
+     * and delegates the command to the resulting protocol adapter instance.
+     *
+     * @param consumerRecord The consumer record corresponding to the command.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public void mapAndDelegateIncomingCommandMessage(final KafkaConsumerRecord<String, Buffer> consumerRecord) {
+        Objects.requireNonNull(consumerRecord);
+
+        final KafkaBasedCommand command;
+        try {
+            command = KafkaBasedCommand.from(consumerRecord);
+        } catch (final IllegalArgumentException exception) {
+            log.debug("command record is invalid", exception);
+            return;
+        }
+
+        final SpanContext spanContext = KafkaTracingHelper.extractSpanContext(tracer, consumerRecord);
+        final Span currentSpan = CommandConsumer.createSpan("map and delegate command", command.getTenant(),
+                command.getDeviceId(), null, tracer, spanContext);
+        final KafkaBasedCommandContext commandContext = new KafkaBasedCommandContext(command, currentSpan);
+
+        command.logToSpan(currentSpan);
+        if (command.isValid()) {
+            log.trace("received a valid command record");
+            mapAndDelegateIncomingCommand(commandContext);
+        } else {
+            log.debug("malformed command message [tenant: {}, device: {}]", command.getTenant(), command.getDeviceId());
```

Comment:
Better to log using `command.toString()` here. That will give details about the reason the command record is invalid:
`log.debug("received invalid command record: {}", command);`.